net/src/ipv4/addr.rs

```diff
@@ -6,6 +6,39 @@
 use std::fmt::{Display, Formatter};
 use std::net::Ipv4Addr;
 
+/// Thin wrapper around [`Ipv4Addr`]
+///
+/// This wrapper scopes addresses to be unicast.
+///
+/// This wrapper is zero cost save for the need to check that the [`Ipv4Addr`] is in fact unicast.
+#[non_exhaustive]
+#[repr(transparent)]
+#[derive(
+    Debug, Clone, Copy, PartialEq, Eq, Hash, PartialOrd, Ord, serde::Deserialize, serde::Serialize,
+)]
+pub struct SourceIpv4Addr(Ipv4Addr);
+
+impl SourceIpv4Addr {
+    /// Map an unqualified [`Ipv4Addr`] to a [`UnicastIpv4Addr`].
+    ///
+    /// # Errors
+    ///
+    /// Returns the supplied address back in the [`Err`] case if it is not a unicast address.
+    pub fn new(ip: Ipv4Addr) -> Result<SourceIpv4Addr, Ipv4Addr> {
+        if ip.is_multicast() || ip.is_unspecified() {
```

Comment:
In theory the unspecified ip address 0.0.0.0 can appear in some packets (e.g. DHCP).
I don't know if forbidding them here may cause trouble depending on the use of `SourceIpv4Addr` ?

Reply:
That is a really good point.

I know that linux  forbids 0.0.0.0 as the source address of a vtep